src/Avalonia.Styling/Styling/NthChildSelector.cs

```diff
@@ -0,0 +1,134 @@
+﻿#nullable enable
+using System;
+using System.Text;
+
+using Avalonia.LogicalTree;
+
+namespace Avalonia.Styling
+{
+    public interface IChildIndexProvider
+    {
+        (int Index, int? TotalCount) GetChildIndex(ILogical child);
```

Comment:
I know you say in the description that this should be changed to an observable, but wouldn't a simple `ChildIndexesChanged` event or something suffice? If an item is removed then all indexes are going to have to be re-evaluated anyway and a single event subscription is going to be better than creating an observable for every item.

Reply:
Yes, event might be better here.

> If an item is removed then all indexes are going to have to be re-evaluated anyway

I was thinking about possible optimization here. Let's image we have 5 items, and for each item we created selector activator that is subscribed on this event. Now, if we remove the last item (5th), we don't need to handle indexes changed event for any other item. Or if we removed only 4th item, then only 5th item should be recalculated. For reversed order it can be similar.
Or probably I was thinking too much about it, so it's not a problem.

Reply:
I guess it depends how expensive the calculation is in the case where nothing has changed.

Reply:
Now I am thinking more about it, and it shouldn't be a problem at all.
Because for huge lists thousands of items we already have virtualization, and styles will be matched only on materialized items.